### DISCUSSION TITLE
Export pure code for deployment

### DISCUSSION THREAD

**Discussion by VaZark** [leon-thomm/Ryven#91] [Ideas]
Hi there! I've not had the chance to test Ryven yet but going through the docs gives me the impression that it is designed to be used as is, locally. Is there any plans to export the code out as python files that can be reused for deployment? 

While Ryven doesn't support import projects, I see this being highly useful as a web dev to track my logic, reuse components and patterns while also being able to show your boss the mess we're actually dealing with

Being able to use this on top of Django or python would be game-changing

**Comment by leon-thomm** [comment]
Hi, and welcome!

The reason I am not particularly looking into parsing existing code is that the graph representation requires conceptually a very different kind of thinking, it's very data processing centric, and also not useful in all situations. When there's a very fixed use case, say you want to turn a single FastAPI or Flask script into a flow, then you could probably write something on top of Ryven that generates the graph, but it depends on the use case and is therefore not included in Ryven itself.

Regarding code export: Ryven simply *runs* code, which is the most general thing you can do. If you want to parse a piece of code out of your nodes, then go ahead and write those nodes so that you can. I have written a post about this a few months ago, see #68. And if you are just looking for executing your Ryven graphs in a non-ui environment, you can do so by deploying directly on ryvencore.

Did I answer your questions? Feel free to comment and ask further questions.

**Comment by VaZark** [comment]
Thanks for the rapid reponse. That was a well thought-out post. It answers my Q perfectly.

However, I stumbled upon Ryven when I was looking Node-based code editors outside of Unreal Engine and Blender to do general programming. 

Looks, I might have to do it myself.

Can you give me your thoughts/take on how difficult the idea can be? I'll be targeting only scripting languages (python/JS).

1. Having a standard intermediate notation for visual programming 
2. Generating the notation by parsing using something like  [Treesitter](https://tree-sitter.github.io/tree-sitter/) **but** limiting to local code and not include the dependencies
3. Generating then Drag and Drop GUI editor for nodes
5. Running code by mocking input for select nodes

**Reply by leon-thomm** [reply]
well, the most effective tech stack depends on your skills and interests. I decided to go for a Qt UI mainly to equally target all desktop platforms simultaneously, without any specializations, because its python bindings provide *by far* the most advanced and scalable UI framework to use in python code, I already had some Qt experience in C++, and because it's fast. But it has also downsides, especially in python.

If you want to target JS, though, or in general potentially web-related content, it might be more sustainable to develop a UI using a JS framework like React, to run in the browser. If you want to expand to desktop there's Electron, and if you then also want to target python you can still write the backend in python and spin up a local server, as you do with Jupyter notebooks, etc. Also notice that there are many JS-based UI solutions for flow-based editors out there, like NodeRED where you might find further inspiration. If I would start a new related project, I would probably go in this direction (which is kinda happening right now :eyes: but I can't say if there will ever be a working prototype).

**Comment by dfpena** [comment]
If you are interested you can use https://pluricorp.com/stremecoder/ it is a flow based python editing software also. It generates pure python code that can be deployed with the only dependency being decorators. The decorators can be removed and the python files made portable and executable using rust ( https://pyoxidizer.readthedocs.io/en/stable/). I do it all the time.
